Agent-based model crashes due to index error

2023-05-10 16:03:13 [INFO] Assigning mode and destination for 1334423 agents (100 % of total population)
2023-05-10 16:07:55 [ERROR] Exception at iteration 1
Traceback (most recent call last):
  File "C:/Users/HajduPe/Documents/helmet-model-system/Scripts/helmet.py", line 104, in main
    else model.run_iteration(impedance, i))
  File "C:\Users\HajduPe\Documents\helmet-model-system\Scripts\modelsystem.py", line 257, in run_iteration
    self._add_internal_demand(previous_iter_impedance, iteration=="last")
  File "C:\Users\HajduPe\Documents\helmet-model-system\Scripts\modelsystem.py", line 571, in _add_internal_demand
    tour.choose_destination(sec_dest_tours)
  File "C:\Users\HajduPe\Documents\helmet-model-system\Scripts\datatypes\tour.py", line 152, in choose_destination
    self.purpose.attracted_tours[self.mode][dest_idx] += 1
IndexError: index 1977 is out of bounds for axis 0 with size 1977
2023-05-10 16:07:55 [ERROR] Fatal error occured, simulation aborted.

Forcing dest_idx to be under 1977 fixes the situation, but we need some more thorough investigation of why this happens.

@zptro
@johpiip

--- a/Scripts/models/logit.py
+++ b/Scripts/models/logit.py
@@ -276,7 +276,7 @@ class ModeDestModel(LogitModel):
             Mode (car/transit/bike/walk) : numpy 2-d matrix
                 Choice probabilities
         """
-        prob = self.calc_basic_prob(impedance)
+        prob = self._calc_prob(self._calc_utils(impedance))
         for mod_mode in self.mode_choice_param:
             for i in self.mode_choice_param[mod_mode]["individual_dummy"]:
                 dummy_share = self.zone_data.get_data(
@@ -299,14 +299,12 @@ class ModeDestModel(LogitModel):
             Mode (car/transit/bike/walk) : dict
                 Type (time/cost/dist) : numpy 2-d matrix
                     Impedances
-        
-        Returns
-        -------
-        dict
-            Mode (car/transit/bike/walk) : numpy 2-d matrix
-                Choice probabilities
         """
-        return self._calc_prob(self._calc_utils(impedance))
+        self._calc_utils(impedance)
+        self.cumul_dest_prob = {}
+        for mode in self.mode_choice_param:
+            cumsum = self.dest_exps[mode].T.cumsum(axis=0)
+            self.cumul_dest_prob[mode] = cumsum / cumsum[-1]
     
     def calc_individual_prob(self, mod_mode, dummy):
         """Calculate matrix of probabilities with individual dummies.
@@ -413,14 +411,11 @@ class ModeDestModel(LogitModel):
 
     def _calc_prob(self, mode_expsum):
         prob = {}
-        self.mode_prob = {}
-        self.cumul_dest_prob = {}
         for mode in self.mode_choice_param:
-            self.mode_prob[mode] = self.mode_exps[mode] / mode_expsum
-            dest_expsum = self.dest_expsums[mode]["logsum"]
-            dest_prob = self.dest_exps[mode].T / dest_expsum
-            prob[mode] = self.mode_prob[mode] * dest_prob
-            self.cumul_dest_prob[mode] = dest_prob.cumsum(axis=0)
+            mode_prob = self.mode_exps[mode] / mode_expsum
+            dest_prob = (self.dest_exps[mode].T
+                         / self.dest_expsums[mode]["logsum"])
+            prob[mode] = mode_prob * dest_prob
         return prob
 
     def _get_cost_util_coefficient(self):

--- a/Scripts/models/logit.py
+++ b/Scripts/models/logit.py
@@ -289,8 +289,9 @@ class ModeDestModel(LogitModel):
         return prob
     
     def calc_basic_prob(self, impedance):
-        """Calculate matrix of mode and destination choice probabilities.
+        """Calculate utilities and cumulative destination choice probabilities.
 
+        Only used in agent simulation.
         Individual dummy variables are not included.
         
         Parameters

